### DISCUSSION TITLE
Is there a way to filter by category: 1320 and assign the normalized string as the folder name during export?

### DISCUSSION THREAD

**Discussion by digikeith** [RhetTbull/osxphotos#1677] [Q&A] [answered]
Hello and thank you for making such a feature-rich interface to Apple Photos. I have read through many of the examples, relevant discussions, and even through the tutorial, but I was not able to find a way to do the following. 

I would like to export my photo library (shared amongst my family) using the **category 1320** and specifically use the **normalized_string** (ex. 'Shared by Dawn) for each folder name, in my case there are 5 people.

For example, I would like to have the following file structure when exporting:  
  
     /path/to/photos/me

     /path/to/photos/dawn

     /path/to/photos/child1

     /path/to/photos/child2

     /path/to/photos/child3

  
I was able to find the info using: **osxphotos inspect** and again using **get_photo('617F4499-ACBA-4FA6-A318-98EA5A1B55C7').search_info.\_db_searchinfo** to find the following from the output:

...{'uuid': '617F4499-ACBA-4FA6-A318-98EA5A1B55C7', 'rowid': 866208, 'uuid_0': -6462741689784828063, 'uuid_1': -4083327409844643677, 'groupid': 38051, **'category': 1320**, 'owning_groupid': 0, 'content_string': 'Shared by Dawn', **'normalized_string': 'shared by dawn'**, 'lookup_identifier': ''},...

so, my question is :

Is there a way to filter by this category and assign the normalized string as the folder name using osxphotos export?

**Comment by RhetTbull** [comment]
Yes, I would use a template function. See the [example here](https://github.com/RhetTbull/osxphotos/blob/main/examples/template_function.py).  The template function gets passed the photo being processed and should return the value to use.  If you get stuck, I'm happy to take a look later this evening.  Basically, get the search_info for the photo passed into the function, find the category (which may or may not be present) and format the value you want then return that or None if not found. Call with `--directory "{function:myfile.py::myfunction}"`

**Reply by digikeith** [reply]
Thank you for the quick reply! I will check out the example templates and see if I can figure it out.

**Comment by digikeith** [comment]
@RhetTbull 

So, I spent some time last night to work on this and I modified the template_function code below trying to pull out the relevant information like you suggested. But, I think I might be missing some steps. Please forgive the script quality, my python skills are rusty.

I call the script using:  

`osxphotos export ~/dev/osxphotos/tmp --directory "{function:template_shared.py::shareDirectory}" --skip-original-if-edited --sidecar XMP --touch-file `

And, I get the following output and error:  (crash log attached)

```
Using osxphotos export database: version 10.0 located at /Users/keith/dev/osxphotos/tmp/.osxphotos_export.db
Processing database /Users/keith/Pictures/Photos Library.photoslibrary/database/photos.db
Processing database /Users/keith/Pictures/Photos Library.photoslibrary/database/Photos.sqlite
Photos database version: 6000, 9.6.
Processing persons in photos.
Processing detected faces in photos.
Processing albums.
Processing keywords.
Processing photo details.
Processing import sessions.
Processing additional photo details.
Processing face details.
Processing photo labels.
Processing EXIF details.
Processing computed aesthetic scores.
Processing comments and likes for shared photos.
Processing moments.
Processing syndication info.
Processing shared iCloud library info
Done processing details from Photos library.
Exporting 56223 photos to /Users/keith/dev/osxphotos/tmp...
Exporting 56223 photos ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% 0:04:37
Something went wrong and osxphotos encountered an error:
argument of type 'NoneType' is not iterable
Crash log written to '/Users/keith/dev/osxphotos/osxphotos_crash.log'
```

Here is the script:

```
""" Example showing how to use a custom function for osxphotos {function} template 
    Use:  osxphotos export /path/to/export --filename "{function:/path/to/template_function.py::example}"

    You may place more than one template function in a single file as each is called by name using the {function:file.py::function_name} format

    (1) get the search_info for the photo passed into the function
    (2) find the category (which may or may not be present)
    (3) format the value you want then return that or None if not found. 
    
    Call with --directory "{function:template_shared.py::shareDirectory}"
    
    """

import pathlib
from typing import List, Optional, Union

from osxphotos import PhotoInfo
from osxphotos import SearchInfo
from osxphotos.phototemplate import RenderOptions

def shareDirectory(
    photo: SearchInfo, options: RenderOptions, args: Optional[str] = None, **kwargs
) -> Union[List, str]:
    """example function for {function} template; adds suffix of # if photo has adjustments and ! if photo is a favorite

    Args:
        photo: osxphotos.PhotoInfo object
        options: osxphotos.phototemplate.RenderOptions object
        args: optional str of arguments passed to template function
        **kwargs: not currently used, placeholder to keep functions compatible with possible changes to {function}

    Returns:
        str or list of str of values that should be substituted for the {function} template
    """
    sharedDir = photo.search_info._db_searchinfo
    
    if 'category' in sharedDir and sharedDir['category'] == 1320:
        normalized_string = sharedDir.get('normalized_string')
        if content_string := sharedDir.get('content_string'):
            if 'Shared by ' in content_string:
                shared_by_words = content_string.split('Shared by ')[1].split()

            return [normalized_string, *shared_by_words]
    
    # If no suitable data was found, just return 
    return "uncategorized"
```

[osxphotos_crash.log](https://github.com/user-attachments/files/16887803/osxphotos_crash.log)

**Reply by RhetTbull** [reply]
@digikeith you were close. I made a few tweaks below. FYI, for testing this, you can use `osxphotos template` which will render a template on the currently selected photo.  (And if you change the selection and re-run the template, it will use the new photo). Type the template string into the prompt after it loads, e.g. `{function:template_shared.py::shareDirectory}` then hit return. You can use up arrow to cycle through previous templates.  This is easier than trying the export command.

The main thing I changed is that `_db_searchinfo` is a list of dicts so you need to cycle through each dict looking at the category. Note that any variable starting with `_` is considered by OSXPhotos to be private and could change and thus break this in the future. In reality, I do not anticipate changes but just FYI. I am going to open a new feature to allow this to be accessed directly from the `search_info` property.

```python
""" Example showing how to use a custom function for osxphotos {function} template
    Use:  osxphotos export /path/to/export --filename "{function:/path/to/template_function.py::example}"

    You may place more than one template function in a single file as each is called by name using the {function:file.py::function_name} format

    (1) get the search_info for the photo passed into the function
    (2) find the category (which may or may not be present)
    (3) format the value you want then return that or None if not found.

    Call with --directory "{function:template_shared.py::shareDirectory}"

    """

import pathlib
from typing import List, Optional, Union

from osxphotos import PhotoInfo
from osxphotos import SearchInfo
from osxphotos.phototemplate import RenderOptions


def shareDirectory(
    photo: SearchInfo, options: RenderOptions, args: Optional[str] = None, **kwargs
) -> Union[List, str]:
    """example function for {function} template; adds suffix of # if photo has adjustments and ! if photo is a favorite

    Args:
        photo: osxphotos.PhotoInfo object
        options: osxphotos.phototemplate.RenderOptions object
        args: optional str of arguments passed to template function
        **kwargs: not currently used, placeholder to keep functions compatible with possible changes to {function}

    Returns:
        str or list of str of values that should be substituted for the {function} template
    """
    sharedDir = photo.search_info._db_searchinfo

    for item in sharedDir:
        if "category" in item and item["category"] == 1320:
            normalized_string = item.get("normalized_string")
            if normalized_string := item.get("normalized_string"):
                if "shared by " in normalized_string:
                    normalized_string = normalized_string.split("shared by ")[1].strip()
                return [normalized_string]

    # If no suitable data was found, just return
    return "uncategorized"
```

**Comment by digikeith** [comment]
It works! I tested it with _osxphotos template_, like you suggested and it looks like it is able to handle each of the use cases I tried.  I will test it out with a full export overnight and see how it goes.

Thank you so much for helping me with this!  I think this is a great option for families that are contributing to a combined icloud photos album, but don't have the ability to see when individual famliy members contribute to the shared library. In my case, this allows me to export directly to immich which gives me more options and control.

**Reply by digikeith** [reply]
Well, I guess my manual testing was not thorough enough. It did crash again with the same ''NoneType' object is not iterable' error. Is it possible that there is a case that is not accounted for? 

Here is how I ran it: 
`osxphotos export /Volumes/data/icloud_photos --skip-original-if-edited --sidecar XMP  --directory "{function:template_shared.py::shareDirectory}" --touch-file  `

Here is the crash log:
[osxphotos_crash.log](https://github.com/user-attachments/files/16900720/osxphotos_crash.log)

**Reply by RhetTbull** [reply, marked as answer]
Try this version.  Apparently `_db_searchinfo` can be None (that is the photo had no categories assigned).

```python
""" Example showing how to use a custom function for osxphotos {function} template
    Use:  osxphotos export /path/to/export --filename "{function:/path/to/template_function.py::example}"

    You may place more than one template function in a single file as each is called by name using the {function:file.py::function_name} format

    (1) get the search_info for the photo passed into the function
    (2) find the category (which may or may not be present)
    (3) format the value you want then return that or None if not found.

    Call with --directory "{function:template_shared.py::shareDirectory}"

    """

import pathlib
from typing import List, Optional, Union

from osxphotos import PhotoInfo
from osxphotos import SearchInfo
from osxphotos.phototemplate import RenderOptions


def shareDirectory(
    photo: SearchInfo, options: RenderOptions, args: Optional[str] = None, **kwargs
) -> Union[List, str]:
    """example function for {function} template; adds suffix of # if photo has adjustments and ! if photo is a favorite

    Args:
        photo: osxphotos.PhotoInfo object
        options: osxphotos.phototemplate.RenderOptions object
        args: optional str of arguments passed to template function
        **kwargs: not currently used, placeholder to keep functions compatible with possible changes to {function}

    Returns:
        str or list of str of values that should be substituted for the {function} template
    """
    sharedDir = photo.search_info._db_searchinfo

    if not sharedDir:
        return "uncategorized"

    for item in sharedDir:
        if "category" in item and item["category"] == 1320:
            if normalized_string := item.get("normalized_string"):
                if "shared by " in normalized_string:
                    normalized_string = normalized_string.split("shared by ")[1].strip()
                return [normalized_string]

    # If no suitable data was found, just return
    return "uncategorized"

```

**Reply by digikeith** [reply]
That was it! To test it out, I pushed the items that did not have _db_searchinfo to a separate directory and only a couple sidecar .xmp files (no images) were being exported there. Let me know if you need more info for that.

Thanks again for your help! I'm going to let the export run through the day while at work and will post the results.
